src/internal_events/throttle.rs

```diff
@@ -4,27 +4,21 @@ use vector_lib::internal_event::{ComponentEventsDropped, InternalEvent, INTENTIO
 #[derive(Debug)]
 pub(crate) struct ThrottleEventDiscarded {
     pub key: String,
+    pub emit_events_discarded_per_key: bool,
 }
 
 impl InternalEvent for ThrottleEventDiscarded {
     fn emit(self) {
-        // TODO: Technically, the Component Specification states that the discarded events metric
-        // must _only_ have the `intentional` tag, in addition to the core tags like
-        // `component_kind`, etc, and nothing else.
-        //
-        // That doesn't give us the leeway to specify which throttle bucket the events are being
-        // discarded for... but including the key/bucket as a tag does seem useful and so I wonder
-        // if we should change the specification wording? Sort of a similar situation to the
-        // `error_code` tag for the component errors metric, where it's meant to be optional and
-        // only specified when relevant.
-        counter!(
-            "events_discarded_total", 1,
-            "key" => self.key,
-        ); // Deprecated.
+        let message = "Rate limit exceeded.";
+
+        debug!(message, key = self.key, internal_log_rate_limit = true);
+        if self.emit_events_discarded_per_key {
+            counter!("events_discarded_total", 1, "key" => self.key);
```

Comment:
Outside of the scope of this PR, but we may want to think about renaming this to `component_events_discarded_total` in the future.

Reply:
Yeah, I'm going back and forth on that. That would require extending the current component specification for the `component_events_discarded_total` metric and could also significantly increase the cardinality of that metric. If there are cardinality issues, I would prefer that the impact be isolated to a one-off, single-use metric rather than impacting a metric used by all components. However, that does result in an inferior UX on the other hand in my opinion.

Reply:
I'll add back the TODO.